src/adapters/surrealDB/index.ts

```diff
@@ -0,0 +1,189 @@
+			const value = payload[key];
+
+			if (role.cardinality === 'ONE' && Array.isArray(value)) {
+				payload[key] = value[0];
+			}
+		}
+	});
+};
+
+const buildQuery = (thing: string, query: EnrichedBqlQuery) => {
+	const attributes = query['$fields'].filter((q): q is EnrichedBqlQueryAttribute => q['$thingType'] === 'attribute');
+	const entities = query['$fields'].filter((q): q is EnrichedBqlQueryEntity => q['$thingType'] === 'entity');
+	const relations = query['$fields'].filter((q): q is EnrichedBqlQueryEntity => q['$thingType'] === 'relation');
+
+	const relationsQuery = relations.map((relation) => {
+		return `(SELECT VALUE meta::id(id) as id FROM <-${relation.$thing}_${relation.$plays}<-${relation.$thing}) as \`${relation.$as}\``;
+	});
+
+	const entitiesQuery = entities.map((entity) => {
+		const role = pascal(entity['$playedBy'].relation);
+
+		return query.$thingType === 'relation'
+			? `(SELECT VALUE meta::id(id) as id FROM ->${role}_${entity['$playedBy']['plays']}.out) as ${entity['$as']}`
+			: `(SELECT VALUE meta::id(id) as id FROM <-${role}_${entity['$playedBy']['path']}<-${role}->${role}_${entity['$playedBy']['plays']}.out) as ${entity['$as']}`;
+	});
+
+	const filterExpr = query['$filter']
+		? `WHERE ${Object.entries(query['$filter'])
+				.map(([key, value]) => `${key} = ${query['$thing']}:\`${value}\``)
+				.join(',')}`
+		: '';
+
+	const result = `SELECT 
+    ${[...attributes.map(convertEntityId), ...entitiesQuery, ...relationsQuery].join(',')}
+  FROM ${thing} 
+  ${filterExpr}
+  FETCH ${entities.map((entity) => entity['$path']).join(',')} 
+  PARALLEL`;
+
+	return result;
+};
+
+const buildSurrealDbQuery: PipelineOperation<SurrealDbResponse> = async (req, res) => {
+	const { dbHandles, enrichedBqlQuery, schema } = req;
+
+	if (!enrichedBqlQuery) {
+		throw new Error('BQL request not parsed');
+	}
+	if (!dbHandles.surrealDB) {
+		throw new Error('missing SurrealDB in dbHandles');
+	}
+
+	const connector = req.config.dbConnectors.find((connector) => connector.provider === 'surrealDB');
+
+	if (!connector) {
+		throw new Error('missing SurrealDB config');
+	}
+
+	const mapItem = dbHandles.surrealDB.get(connector.id);
+
+	if (!mapItem) {
+		throw new Error(`missing SurrealDB client with id of ${connector.id}`);
+	}
+
+	const { client } = mapItem;
+
+	const results = await Promise.all(
+		(req.enrichedBqlQuery as Array<EnrichedBqlQuery>).map(async (query, idx) => {
+			let queryResult: Array<Record<string, unknown>> | undefined;
+
+			if (query['$thingType'] !== 'entity') {
+				// NOTE relations from testSchema contains hyphen at the moment. Remove case transformation, once we use camelcase for all tables
+				const thing = pascal(query.$thing);
+
+				const generatedQuery = buildQuery(thing, query);
+
+				const queryRes: Array<
+					Record<string, unknown> & {
+						id: string;
+					}
+				> = await client.query(generatedQuery);
+
+				queryResult = queryRes.map((item) => ({
```

Comment:
How do you manage tree structures? Or is still to-do?

Reply:
That is still todo, not implemented at the moment.